From 2370ade20f24f5be3248be561b83e77c3acb3f89 Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Wed, 1 Nov 2017 19:09:12 +0100
Subject: [PATCH 4/4] scsi: sr: wait for the medium to become ready.

Patch-mainline: no, kabi
References: bsc#1048585

When opening the cd-rom determine disk status. As a sideeffect the
logic that waits for the disk to become ready is invoked.

NOTE: this is alternative to changing the cdrom.c interface to be able
to do this in driver-neutral way.

Signed-off-by: Michal Suchanek <msuchanek@suse.de>
---
 drivers/scsi/sr.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/drivers/scsi/sr.c
+++ b/drivers/scsi/sr.c
@@ -533,6 +533,16 @@ static int sr_block_open(struct block_de
 
 	mutex_lock(&sr_mutex);
 	ret = cdrom_open(&cd->cdi, bdev, mode);
+	/* wait for drive to get ready */
+	if ((ret == -ENOMEDIUM) && !(mode & FMODE_NDELAY))
+		switch (sr_disk_status(&cd->cdi)) {
+			case CDS_NO_DISC:
+			case CDS_NO_INFO:
+			case CDS_AUDIO:
+				break;
+			default: /* looks like data disc was detected */
+				ret = cdrom_open(&cd->cdi, bdev, mode);
+		}
 	mutex_unlock(&sr_mutex);
 
 	scsi_autopm_put_device(sdev);
